### ISSUE TITLE
2.2.1

### ISSUE THREAD

**Issue by 1to9only** [1to9only/HoDoKu-Discuss#3]
I am comparing CCV54's HoDoKu_221.jar against hobiwans's 2.2.0 hodoku.jar (released 2012-07-29).
The intention in to be able to build CCV54's 2.2.1 hodoku.jar from java source code.
The code/file changes from hobiwans's 2.2.0 to CCV54's 2.2.1 are noted here.

[Edit]
The reference CCV54's 2.2.1 hodoku.jar I've been using is the one posted on yoki123/hodoku here:
https://github.com/yoki123/hodoku/issues/1#issuecomment-684437806
HoDoKu_221.zip, 1,511,602 bytes
HoDoKu_221.jar, 1,659,252 bytes, 17-10-19 11:52:04

**Comment by 1to9only** [comment]
**Folder**: img - files below to be included in 2.2.1

**files differ**:
-f_1c.png
-f_1c_inactive.png
-f_2c.png
-f_2c_inactive.png
-f_3c.png
-f_3c_inactive.png
-f_4c.png
-f_4c_inactive.png
-f_5c.png
-f_5c_inactive.png
-f_6c.png
-f_6c_inactive.png
-f_7c.png
-f_7c_inactive.png
-f_8c.png
-f_8c_inactive.png
-f_9c.png
-f_9c_inactive.png
-f_xyc.png
-ov078.png
-rgDeselected1.png
-rgSelected1.png

**new files**:
-f_xyc_inactive.png
-inactive.png

**Comment by 1to9only** [comment]
**Typo fix**:
**File**: src\intl\MainFrame.properties
**2.2.0**: MainFrame.vageHintToolButton.toolTipText=Show vage hint
**2.2.1**: MainFrame.vageHintToolButton.toolTipText=Show vague hint

**Comment by CCV54** [comment]
> **Folder**: img - files below to be included in 2.2.1

> **new files**:
> f_xyc_inactive.png
> inactive.png
> 
> [Edit] I am still a few weeks away from even looking at the above image files, but I'll likely select the images as follows:
> 
> * use image from hobiwan's HoDoKu on sourceforge, [r147] last version.
> * replace with 2.2.1 image (listed above, from HoDoKu_221.jar) if different.
> * replace image if the one in sudoku_filter.zip (from #1) is different.
> 
> @CCV54, you may post newer (or other) images for inclusion in 2.2.1 if you wish.

@1to9only, the files in sudoku_filter.zip _are_ different to those in HoDoKu_221.jar. Very subtly so (added 5% transparency), but, in my view, it does make a difference.

f_xyc_inactive.png is not required. However, inactive.png **is** for the inactive state on ColorKu Filter bar! Come to think of it, there must be some code changes in relation to that as well.

Have been considering editing both rgDeselected1.png and rgSelected1.png, but, at this stage, not sure it would be any great advantage.
Oh, and.. Ahaha, glad you noticed the spelling change for tooltip for on vague hint. Only remembered that, like, yesterday.

Please, take all the time you want or need. I'm just happy to have your assistance at all.

**Comment by 1to9only** [comment]
For the 2.2.1 release, I'll also change the About box to reflect the project changes.
![new_hodoku_about](https://user-images.githubusercontent.com/53971101/93759553-51290b80-fc02-11ea-8e25-e826f219b2b0.png)

**Comment by 1to9only** [comment]
2 easy fixes are done:
- fix buttons 3/4/5 selection from ColorKu color panel
- fix typo in HintToolButton.toolTipText

**Comment by 1to9only** [comment]
Also uploaded images from @CCV54's HoDoKu_221.jar - this is referenced in 1st post above.
I want to build the equivalent HoDoKu_221.jar before making any new image changes.
The same will apply for any Color changes, so I'll be using HoDoKu_221.jar Colors.

**Comment by CCV54** [comment]
> For the 2.2.1 release, I'll also change the About box to reflect the project changes.
> @CCV54
> Do you want to be given credit for the post 2.2.0 changes? I'll prob change the About box again, when it's time to update it, just prior to setting the version to 2.2.1, then build and release 2.2.1 Hodoku.jar. Something like this:
> 
> ```
>  Copyright (c) 2008-12 Bernhard Hobiger
> https://sourceforge.net/projects/hodoku/
>  Copyright (c) 2018-20 CCV54, 1to9only
>   https://github.com/1to9only/HoDoKu
> ```

Hmm.. I'm already given credit, as "ccv", for some contributions to 2.2.0. That was for work on ColorKu mask, colours and colour changes more broadly. I don't see that my contributions now go very much further, just refinements on what I already did. Did offer an earlier version of new filter numbers too. Not sure why they weren't accepted. All I know is that Bernhard reported being hospitalised at one time and otherwise unwell and lacking time to commit to the project. I ended up in hospital too, so that some of my intended changes never made it into 2.2.0. Iirc, "ddyer" contributed, I'm guessing his, ColorKu setup or idea in the first place.
Btw: Hobiwan was satisfied with the ColorKu mask which made it into 2.2.0. Me, not so much. It never did look quite as good as what I had produced in my lab. Turns out it was not properly aligned, and actually the wrong size, which produced some ugly effects. Managed to fix that while reworking the mask, which I had always in mind to do. I guess if copyright belongs to me at all, then it is mostly for image work. Otoh, getting nine colours to play nice together and work as well possible with the mask was no little effort. So....

The confirmation dialogue idea was accepted by hobiwan, but for application in the next release. Probably a few other things as well. The only thing I can claim to be my idea is Concrete hint for XYZ-Wing displaying 3 candidates. Pretty minor changes, in my view, and no more deserving of copyright claim than anything I did earlier. Although, in relation to the programmers I hired, I do technically own the copyright. If anything, I'd prefer simply CCV as that is my tag on the sourceforge board - and anywhere else it is accepted. :)

P.S. There's been a few requests for users to be able draw their own lines, for various Chain techniques, and, again, accepted by hobiwan as a possibility for a future release. Not sure he made any progress toward that. Maybe you can tell. ? Anyway, cell or candidate colouring, I reckon, will find the links just as well.

**Comment by 1to9only** [comment]
Just want to get from A to B (2.2.0->2.2.1) using hobiwan's 2.2.0 and your 2.2.1 as reference points, before going to C (2.2.2).
I've noted the changes you've asked, they'll all go into 2.2.2. There'll be an audit trail of changes, and from the project's point it's easier to manage for me. Hope you'll see the wisdom of this later!

**Comment by 1to9only** [comment]
I've now added the default colors from HoDoKu_221.jar, 1,659,252 bytes, 17-10-19 11:52:04 (those that are changed from 2.2.0 to 2.2.1). To see what's changed, go here: https://github.com/1to9only/HoDoKu/commit/a4dceb430f14212e1961027653188c5767c6095f
The removed 2.2.0 lines are in light pink, the new updated lines are in light green.

**Comment by CCV54** [comment]
@1to9only Now I remember that a comma is used as a separator between R:G:B values in the code itself. Looks like comma and space, actually. Is that correct?
I can edit my most recent colour list to reflect that, if it would make it easier for you. I.e., copy&paste?

Btw, there's more than the 9 colours for ColorKu involved in my changes (you already know). That's because I need to check how well or not the other 20 or so colours work with each other, and in combination in both Sudoku and ColorKu. It's maybe not as tedious and time consuming as the coding I'm now asking for, but enough that I think I've earned some credit for the latest manifestation of 2.2.1. The most recent image files didn't emerge overnight either. So, yeh, copyright 2018-2020 including CCV54, thanks.

Haven't heard a peep re the correct source code. Not expecting to really, but...

**Comment by 1to9only** [comment]
I do other things (to be expected!), so only spend short periods of time each day on HoDoKu - at this rate, 2.2.1 will be available early next week.

Yes, RGB colors are coded like 'Color(255, 188,  81)' in the code. The R:G:B is not an issue, as I'll copy&paste the whole lot to a text file, replace ':' with ', ', and work from there. Not looked at the new chages yet, but if I need clarification, I'll ask.

**Comment by 1to9only** [comment]
I've now uploaded code for:
- fix hint for three candidates XYZ-Wing
- add ColorKu inactive image filter

Only the Confirm dialog left to do. I'll create a pre-release for you to test that everything is as your own 2.2.1 version, before finishing the About box and doing the version and build numbers.

**Comment by 1to9only** [comment]
```
Values: 0:102:102
```
I have this mapped to:
```
CANDIDATE_COLOR = new Color (100, 100, 100);     // correct candidates
```
```
Normal: 245:245:255
Altern.: 245:245:255
Selection: 255:255:150
Forbidden: 242:188:185
Possible: 200:221:227
```
I have these mapped to:
```
DEFAULT_CELL_COLOR = Color.WHITE;                // Background normal cell
ALTERNATE_CELL_COLOR = Color.WHITE;              // Background normal cell in every other block
AKT_CELL_COLOR = new Color (255, 255, 150);      // Background currently selected cell
INVALID_CELL_COLOR = new Color (242, 188, 185);  // Background cell with invalid value
POSSIBLE_CELL_COLOR = new Color (207, 239, 232); // Background cell with possible value
```

Can u confirm?

**Comment by 1to9only** [comment]
In PseudoFish's Hodoku, these are his new FILTER constants:
```
public static final Color INVERSE_FILTER_COLOR = new Color(255, 185, 185);
public static final Color FILTER_COLOR = new Color(185, 255, 185);
public static final Color FILTER_GIVEN_CELL_COLOR = new Color(108, 216, 108);
```
assigned to these variables to be used in the program:
```
private Color invalidCellColor = INVERSE_FILTER_COLOR;
private Color possibleCellColor = FILTER_COLOR;
private Color possibleFixedCellColor = FILTER_GIVEN_CELL_COLOR;
```

Apologies about the programming speak!

In my current Hodoku, these are the constants:
```
public static final Color INVALID_CELL_COLOR = new Color(242, 188, 185);
public static final Color POSSIBLE_CELL_COLOR = new Color(207, 239, 232);
```
assigned to these variables for use in the program:
```
private Color invalidCellColor = INVALID_CELL_COLOR;
private Color possibleCellColor = POSSIBLE_CELL_COLOR;
```

The INVALID_CELL_COLOR and POSSIBLE_CELL_COLOR are already noted in the previous post.

I think you're then asking for an implementation of this:
```
possibleFixedCellColor = POSSIBLE_FIXED_CELL_COLOR; [i.e. FILTER_GIVEN_CELL_COLOR]
```

There should be some code that goes with the above, that will have to be borrowed from PseudoFish's Hodoku.

Is the RGB value OK with you?
```
POSSIBLE_FIXED_CELL_COLOR = new Color(108, 216, 108);
```

**Comment by CCV54** [comment]
@1to9only I don't have a clue what this line means:

CANDIDATE_COLOR = new Color (100, 100, 100);     // correct candidates
100, 100, 100 is WHITE anyway...

As for Candidate and Value colours, all I can find in https://github.com/1to9only/HoDoKu/commit/ae5c9b54fa567b59c3d550db7616fa10e8118e0d is:
Lines 400 and something
private Color cellValueColor = CELL_VALUE_COLOR;
private Color candidateColor = CANDIDATE_COLOR;

The Candidate colour, I left alone. The _Value_ colour was (is by default) 0, 0, 255. I want it to be 0, 102, 102.

With some corrections, I can confirm these however:

DEFAULT_CELL_COLOR = Color (245, 245, 255);                // Background normal cell
ALTERNATE_CELL_COLOR = Color (245, 245, 255);              // Background normal cell in every other block
AKT_CELL_COLOR = new Color (255, 255, 150);      // Background currently selected cell
INVALID_CELL_COLOR = new Color (242, 188, 185);  // Background cell with invalid value
POSSIBLE_CELL_COLOR = new Color (200,221, 227); // Background cell with possible value

Yes, code speak is mostly beyond me. I don't know what the difference is between "public" and "private entries, nor what you might be able to do with such "private" code.

And, yes please, _all_ the otherwise listed colour changes packed into 2.2.2 - as changes from 2.2.1. Basically the major point of this exercise. Some I haven't changed, that's true, but I don't remember which.

Congrats on getting the XYZ-Wing and ColorKu filter things done! Many thanks, again!

**Comment by CCV54** [comment]
@1to9only another subtle change to filter numbers. Not sure how long it takes for me to conclude I've done my best. ;)
[sudoku_f_active.zip](https://github.com/1to9only/HoDoKu-Discuss/files/5274207/sudoku_f_active.zip)

**Comment by 1to9only** [comment]
I've uploaded the Commit dialog stuff, and built a pre-release 2.2.1 for you to run a quick test.
Download from: https://github.com/1to9only/HoDoKu/releases

The Hodoku.jar is equivalent to your HoDoKu_221.jar with exceptions:
- there are 2 'vague' fixes
- I've centered the Confirm dialog on Hodoku, instead of on Desktop.

The About... box is also what it will look like - I've made the Build number use the Git Commit numbers - This is a manual change along with version number that will be done for each release.

If all is well, the official 2.2.1 will be released next week.

**Comment by 1to9only** [comment]
@CCV54, Which java version are you using? In a Command Prompt, enter:
```
java.exe -version
```
and post the results here.

Do (Can) you test your images changes?
E.g. Unzip Hodoku.jar, replace images, repack the jar file.
For repacking the jar file, you'll need the jdk (java development kit) and not the jre (java runtime environment).
I am running 'java version "1.8.0_261"'.

**Comment by 1to9only** [comment]
From using Hodoku a little bit more: Pretty much all the colours (those editable in Edit->Preferences anyways) can be customised using the hodoku.hcfg file? Have you explored this method?

**Comment by CCV54** [comment]
@1to9only java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b12, mixed mode)

Image changes I do test, yes. I wasn't aware of any need to "repack" the jar file in that case. It _seems_ to work well enough simply opening the archive and replacing image files.
And, yes, pretty much all the colours, and quite a few other options are customisable and all preferences/changes can be saved together in a configuration file. When I posted 2.2.1 on sourceforge HoDoku discussion I posted config files as well, to account for colour changes I made at the time - actually two at a time, in an effort to allow for some differences in brightness:contrast settings.
Personally, I have one config for standard sudoku and another for ColorKu mostly because Candidate factor needs to be a little different for my purposes. I'm not so sure such changes can be applied except by _loading_ saved config file(s). After that tho, HoDoKu does save current config upon closure and reopens with the same.

Seems to me, especially given the time I've spent refining the whole colour scheme and particularly ColorKu, it would be ideal to present the program with my current colours set as default.

Re, AKT_CELL_COLOR = new Color (255, 255, 150); // Background currently selected cell:
That relates to the "active" (currently selected) cell in Active Cell view as well. In UI Preferences > Colors it is labelled "Selection:" = "Auswahl:" in German. Might make more sense if it was labelled Selected: and the German equivalent Ausgewählt:. 
However, for the sake of consistency, I'd suggest changing it to Active Cell: and Aktive Zelle: or, depending on space allowed, Active: and Aktive:. Seems to me, the code "AKT" is obviously an abbreviation of Aktive anyway.
Bernhard originally had it labelled Cursor:, which was confusing to me. After some discussion with me, he changed it to Selection:.

P.S. The developers I hired are telling me now "we are checking our system if we still have the back up of it [the source code] so please give us some time".......

**Comment by 1to9only** [comment]
Not sure if you have the jdk or jre. Not necessary if your method of testing images works, no need to change.

I'll look to do the AKT change in 2.2.2. I've copied the text over to #4 .

If the developers send the source code soon, post them here - only for me to validate my changes. But not necessary, as I've compared the GitHub 2.2.1 and 2.2.0 sources, the changes are similar to the differences between hobiwan's 2.2.0 and your 2.2.1.

In 2.2.2, I can redirect Help > Report error... and Ask a question... to https://github.com/1to9only/HoDoKu/issues - but in https://github.com/1to9only/HoDoKu/issues/1 I've said that I'm unlikely to be fixing many/any of them!

**Comment by 1to9only** [comment]
Over at forum.enjoysudoku.com: http://forum.enjoysudoku.com/yzf-sudoku-t36846.html
yzfwsf has produced a Hodoku clone rewritten in C# with some newer solving techniques like Exocets, there's no ColorKu.
The most recent version (612) was posted yesterday here: http://forum.enjoysudoku.com/yzf-sudoku-t36846-165.html#p295445

**Comment by CCV54** [comment]
@1to9only Yeh, well.. I don't know the first thing about any possible additional solving techniques. I get the impression that hobiwan was working toward something of the sort - [r147] Tabling new, work in progress... Others have suggested new techniques or reported 'problems' with the solver engine as it stands in 2.2.0. As far as that goes, all I might ask is that you have a look at what hobiwan has done in r147 to see if any of it might be implemented sufficiently to be included here. For another rainy day, perhaps?

As for JRE vs. JDK, I do have JDK 10 (for some reason?) and JDK 8 because I was trying to use JDK to "extract" java files to see if I could do anything toward the changes I wanted. Gave up on that idea pretty quickly. However, my jar files open with "Java(TM) Platform SE binary". I'm assuming that's JRE.
Btw, is there any chance of you packaging this as an exe as well - once complete? The exe being "recommended for Windows".
I have done it before, and guess I could do it again, but...

Re the About info:
I'd suggest arranging it like this, to make it clear that the acknowledgements are from Bernhard Hobiger and not from us.
![About](https://user-images.githubusercontent.com/26049501/94321714-c0707800-ffd3-11ea-8a36-18c347fc8d3e.png)

**Comment by 1to9only** [comment]
hobiwan source has a launch4j folder, which I think was used to package the jar into an exe. I'll investigate, don't foresee any problem creating the exe. I'll think about the installer, it's really for install/uninstall and having icons on the Start Menu, Desktop, etc.

I'm OK with the About box change, might remove a blank line so hobiwan's copyright and thanks are in one block.

**Comment by CCV54** [comment]
Oh.. launch4j, yes, that would do it. Always wondered what he did use. I got NetBeans as well, for a rainy day. Not much point me even trying while dissatisfied as I was with ColorKu and unable, on my own, to set it right. So, many thanks again!

With fresh eyes this morning, I find another desired colour change on top of yesterday's. Pen A-1 255:187:59
It's never going to be "perfect", but I have to keep fiddling until I'm satisfied it's the best I can do.

Btw, you described XYZ-Wing fix as "easy", I notice. The 'professionals' I hired had a real struggle with that one. Ha!
